src/v/serde/parquet/metadata.cc

```diff
@@ -0,0 +1,954 @@
+    IntType be_form = ss::cpu_to_be(v);
+    // NOLINTNEXTLINE(*reinterpret-cast*)
+    buf.append(reinterpret_cast<const char*>(&be_form), sizeof(v));
+}
+
+template<typename IntType>
+void write_uvint(iobuf& buf, IntType v) {
+    bytes b = unsigned_vint::to_bytes(v);
+    buf.append(b.data(), b.size());
+}
+
+enum class field_type : uint8_t {
+    boolean_true = 1,
+    boolean_false = 2,
+    i8 = 3,
+    i16 = 4,
+    i32 = 5,
+    i64 = 6,
+    f64 = 7,
+    binary = 8,
+    list = 9,
+    set = 10,
+    map = 11,
+    structure = 12,
+    uuid = 13,
+};
+
+// A Struct is a sequence of zero or more fields, followed by a stop field. Each
+// field starts with a field header and is followed by the encoded field value.
+// The encoding can be summarized by the following BNF:
+//
+// struct        ::= ( field-header field-value )* stop-field
+// field-header  ::= field-type field-id
+//
+// Compact protocol field header (short form) and field value:
+// +--------+--------+...+--------+
+// |ddddtttt| field value         |
+// +--------+--------+...+--------+
+//
+// Compact protocol field header (1 to 3 bytes, long form) and field value:
+// +--------+--------+...+--------+--------+...+--------+
+// |0000tttt| field id            | field value         |
+// +--------+--------+...+--------+--------+...+--------+
+//
+// Compact protocol stop field:
+// +--------+
+// |00000000|
+// +--------+
+class struct_encoder {
+public:
+    static inline const bytes empty_struct = {0}; // NOLINT
+
+    template<typename T>
+    void write_field(int16_t field_id, field_type type, T val) {
+        auto delta = field_id - _current_field_id;
+        constexpr int16_t max_short_form_delta = 16;
+        if (delta > 0 && delta < max_short_form_delta) {
+            write_short_form_field_header(static_cast<uint8_t>(delta), type);
+        } else {
+            write_long_form_field_header(type, field_id);
+        }
+        _current_field_id = field_id;
+        if constexpr (std::is_same_v<T, iobuf>) {
+            _buf.append(std::move(val));
+        } else {
+            _buf.append(val.data(), val.size());
+        }
+    }
+
+    iobuf write_stop() && {
+        write_be<uint8_t>(_buf, 0);
+        return std::move(_buf);
+    }
+
+private:
+    void write_short_form_field_header(uint8_t delta, field_type type) {
+        auto header_byte = static_cast<uint8_t>(type);
+        header_byte |= static_cast<uint8_t>(delta << 4U);
+        write_be<uint8_t>(_buf, header_byte);
+    }
+
+    void write_long_form_field_header(field_type type, int16_t field_id) {
+        write_be<uint8_t>(_buf, static_cast<uint8_t>(type));
+        write_uvint<int16_t>(_buf, field_id);
+    }
+
+    iobuf _buf;
+    int16_t _current_field_id = 0;
+};
+
+bytes encode_string(std::string_view str) {
+    auto b = unsigned_vint::to_bytes(static_cast<int64_t>(str.size()));
+    b.reserve(b.size() + str.size());
+    for (const char c : str) {
+        b.push_back(c);
+    }
```

Comment:
i wonder if it is better to go directly to iobuf for strings?

Reply:
I expect most strings to fit within the SSO length (this will only be used for field names and metadata), so there will be an advantage of only being a single heap allocation if it fits within SSO (because this always is appended to an existing iobuf).

Reply:
cool makes sense